nft_ingester/src/metadata_workers/downloader.rs

```diff
@@ -0,0 +1,125 @@
+            refresh_metadata_tasks_rx,
+            shutdown_rx,
+            worker_pool,
+        }
+    }
+
+    pub async fn run(self) {
+        let parallel_tasks_for_refresh =
+            self.json_worker.num_of_parallel_workers / PENDING_TASKS_WORKERS_RATIO + 1;
+        let parallel_tasks_for_pending =
+            self.json_worker.num_of_parallel_workers - parallel_tasks_for_refresh + 1;
+
+        tokio::join!(
+            self.download_tasks(parallel_tasks_for_pending, TaskType::Pending),
+            self.download_tasks(parallel_tasks_for_refresh, TaskType::Refresh),
+        );
+    }
+
+    pub async fn download_tasks(&self, parallel_tasks: i32, task_type: TaskType) {
+        let task_receiver = match task_type {
+            TaskType::Pending => self.pending_metadata_tasks_rx.clone(),
+            TaskType::Refresh => self.refresh_metadata_tasks_rx.clone(),
+        };
+
+        for _ in 0..parallel_tasks {
+            let mut shutdown_rx = self.shutdown_rx.resubscribe();
+            let task_receiver = task_receiver.clone();
+            let metadata_to_persist_tx = self.metadata_to_persist_tx.clone();
+            let json_worker = self.json_worker.clone();
+            self.worker_pool.lock().await.spawn(async move {
+                tokio::select! {
+                    _ = shutdown_rx.recv() => {
+                        debug!("Shutting down metadata downloader worker");
+                    },
+                    _ = Self::process_downloading(task_receiver, metadata_to_persist_tx, json_worker) => {},
+                }
+
+                Ok(())
+            });
+        }
+    }
+
+    async fn process_downloading(
+        tasks_receiver: Arc<Mutex<Receiver<MetadataDownloadTask>>>,
+        metadata_to_persist_tx: Sender<(
+            String,
+            Result<MetadataDownloadResult, JsonDownloaderError>,
+        )>,
+        json_worker: Arc<JsonWorker>,
+    ) {
+        loop {
+            match tasks_receiver.lock().await.try_recv() {
+                Ok(task) => {
+                    let begin_processing = Instant::now();
+                    let response = json_worker.download_file(&task, CLIENT_TIMEOUT).await;
+                    json_worker.metrics.set_latency_task_executed(
+                        "json_downloader",
+                        begin_processing.elapsed().as_millis() as f64,
+                    );
+                    if let Err(err) =
+                        metadata_to_persist_tx.send((task.metadata_url, response)).await
+                    {
+                        error!(
+                            "Error during sending JSON download result to the channel: {}",
+                            err.to_string()
+                        );
+                    }
+                },
+                Err(TryRecvError::Disconnected) => {
+                    error!("Could not get JSON task from the channel because it was closed");
+                },
```

Comment:
if the channel is closed we should break out of the loop, it won't be reopened and we'll loop here indefinitely